kratos/elements/distance_calculation_flux_based_element.cpp

```diff
@@ -0,0 +1,502 @@
+
+    //Gauss point container (using nodal integration)
+    const bounded_matrix<double, NumNodes, NumNodes> Ncontainer=IdentityMatrix(NumNodes); 
+
+    //Looping Gauss Points
+    for (unsigned int igauss = 0; igauss<NumNodes; igauss++) {
+        //Getting the correct GP
+        array_1d<double, TNumNodes > N = row(Ncontainer, igauss);
+
+        //Velocity in GP
+        array_1d<double, TDim > vel_gauss = ZeroVector(TDim);
+        for (unsigned int i = 0; i < NumNodes; i++)
+        {
+            for (unsigned int k = 0; k<TDim; k++)
+                vel_gauss[k] += N[i] * v[i][k];
+        }
+
+        const double d_gauss = inner_prod(N,nodal_values);
+
+        if (d_gauss <= 0.0) {
+            vel_gauss = -vel_gauss;
+        }
+
+        const double norm_vel = norm_2(vel_gauss);
+
+        array_1d<double, NumNodes > a_dot_grad = prod(DN_DX_container[igauss], vel_gauss);
+
+        const double tau_denom = std::max(2.0 * norm_vel / h ,  1e-3); 
+        const double tau = 1.0 / (tau_denom);
+
+        const double source_term = d_gauss > 0 ? norm_vel*mCorrectionCoefficient : -norm_vel*mCorrectionCoefficient;
+
+        if(add_convection){
+            //convection + convection stabilization
+            noalias(aux2) += outer_prod(N, a_dot_grad)*gauss_weights[igauss];
+            noalias(aux2) += tau*outer_prod(a_dot_grad, a_dot_grad)*gauss_weights[igauss];
+
+            //source
+            rhs_volumetric_heat += source_term*N*gauss_weights[igauss]; 
+            rhs_volumetric_heat += (tau*source_term)*a_dot_grad*gauss_weights[igauss];
+        } else {
+            noalias(aux2) += tau*prod(DN_DX_container[igauss],trans(DN_DX_container[igauss]))*gauss_weights[igauss];
+        }
+
+        //spherical diffusion to enhance stabilization
+        double spherical_diffusion = 0.1*tau*(norm_vel*norm_vel);
+        noalias(aux2) += spherical_diffusion*prod(DN_DX_container[igauss],trans(DN_DX_container[igauss]))*gauss_weights[igauss];
+    }
+
+    //adding to system LHS
+    noalias(rLeftHandSideMatrix) += aux2;
+
+    //adding to system RHS
+    noalias(rRightHandSideVector) += rhs_volumetric_heat; //external forces
+
+
+    //take out the dirichlet part to finish computing the residual
+    noalias(rRightHandSideVector) -= prod(rLeftHandSideMatrix, nodal_values);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Public Inquiry
+
+template <unsigned int TDim , unsigned int TNumNodes >
+int DistanceCalculationFluxBasedElement<TDim, TNumNodes >::Check(const ProcessInfo &rCurrentProcessInfo) const
+{
+    KRATOS_TRY;
+    // Generic geometry check
+    int out = Element::Check(rCurrentProcessInfo);
+    if (out != 0) {
+        return out;
+    }
+
+    const GeometryType& r_geometry = this->GetGeometry();
+
+    for(unsigned int i=0; i<NumNodes; ++i){
+        const Node<3>& rNode = r_geometry[i];
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISTANCE,rNode);
```

Comment:
You also use `VELOCITY` right?

Reply:
no, it's the POTENTIAL_GRADIENT (non historical)